sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchIterator.scala

```diff
@@ -107,22 +107,30 @@ class GpuColumnarBatchWithPartitionValuesIterator(
 
   private var leftValues: Array[InternalRow] = partValues
   private var leftRowNums: Array[Long] = partRowNums
+  private var nextBatchIter: Iterator[ColumnarBatch] = EmptyGpuColumnarBatchIterator
 
-  override def hasNext: Boolean = inputIter.hasNext
+  override def hasNext: Boolean = {
+    // Check if there is a next batch available, or if we need to process the current batch
+    if (!nextBatchIter.hasNext && inputIter.hasNext) {
```

Comment:
Nit: hasNext might be expensive, would be nice to not call this more than necessary, e.g.:
```
if (nextBatchIter.hasNext) {
  true
} else if (inputIter.hasNext) {
  ...
} else {
  false
}
```